apps/email/js/cards/cmp/autocomplete.js

```diff
@@ -0,0 +1,476 @@
+     */
+    attachedCallback: function() {
+      this.onResize = this.onResize.bind(this);
+      window.addEventListener('resize', this.onResize, false);
+    },
+
+    /**
+     * Custom element callback when element is removed from the document.
+     */
+    detachedCallback: function() {
+      window.removeEventListener('resize', this.onResize, false);
+    },
+
+    /**
+     * Triggered by input focus events.
+     */
+    onFocus: function(event) {
+      this.inputNode = event.target;
+      this.query = this.inputNode.value.trim();
+      this.match();
+    },
+
+    /**
+     * Triggered by input blur events.
+     */
+    onBlur: function(event) {
+      this._blurRequest = Date.now();
+      // Taps on the autocomplete list can trigger a blur, so give those taps
+      // a chance to emit events before hiding the list.
+      setTimeout(() => {
+        // If a match click has tapped while the blur has been initiated,
+        // skip it.
+        if (this._blurRequest > this._matchClickedTime) {
+          this.hide();
+        }
+      }, 200);
+    },
+
+    /**
+     * Triggered by input element input events.
+     * @param  {[type]} event [description]
+     * @return {[type]}       [description]
+     */
+    onInput: function(event) {
+      this.query = this.inputNode.value.trim();
+      this.match();
+    },
+
+    /**
+     * Handles window resizes. Could be triggered by keyboard size changes.
+     */
+    onResize: function(event) {
+      if (this.inputNode && !this.isHidden) {
+        this.sizeList();
+      }
+    },
+
+    /**
+     * Handles taps inside the autocomplete. The tap could occur outside a match
+     * element, in which case means the autocomplete should just close.
+     */
+    onMatchClick: function(event) {
+      var node = event.originalTarget;
+      if (!node.classList.contains('cmp-autocomplete-item')) {
+        while ((node = node.parentNode)) {
+          if (node.classList && node.classList
+                                .contains('cmp-autocomplete-item')) {
+            break;
+          }
+        }
+      }
+
+      if (node && node.match) {
+        this.dispatchEvent(new CustomEvent('autocompleteSelected', {
+          detail: {
+            match: node.match,
+            inputNode: this.inputNode
+          }
+        }));
+        this._matchClickedTime = Date.now();
+      }
+    },
+
+    /**
+     * Loads the dataSource module.
+     * @param  {String} dataSourceId The module ID for the data source.
+     */
+    bindDataSource: function(dataSourceId) {
+      require([dataSourceId], (dataSource) => {
+        this.dataSource = dataSource;
+        if (this.query) {
+          this.match();
+        }
+      });
+    },
+
+    /**
+     * Called to trigger finding matches from the dataSource. Set on a timeout
+     * to avoid multiple fast keystrokes from triggering too many rapid queries.
+     */
+    match: function() {
```

Comment:
naming nit: I'm not crazy about having `match` be a "go do it" function on the custom element and at the same time a result value object of the form { name, address } annotated onto the result-holding inert child elements.  Note that I'm more concerned about the noun/verb dual usage as opposed to the potential for type confusion between the parent and the children.